### PR TITLE
`struct Rav1dIntraPredDspContext`: Fix defaults

### PR DESCRIPTION
**Pull request by kkysen**
This replaces the `Option`s on `.intra_pred` with an `initialized: bool` field as they are always `Some` even if `.is_none()` is checked.

`.is_none()` appears to be checked so that it can lazily initialize DSPs, relying on things being zero-initialized to begin with. 
 Ideally we should move initialization to DSP allocation so its not lazy, but for now (and the way closest to the C), we can use a single initialized flag instead of `Option`s on everything.

This also default initializes `.cfl_pred` as it otherwise isn't fully initialized and is thus leaving uninitialized memory, which is UB if it is made a reference at any point (which it is).  This uses `DefaultValue::DEFAULT` for initialization instead of `None`, though, which moves the check from before the call (`.unwrap()`/`.expect()`) to inside a call that should never happen (`unimplemented!()`).  Normally those `unimplemented!()`s will be dead code eliminated, but here they won't.  Thus, it shouldn't add any perf cost and only a small binary size cost.

### CODE REVIEW

**Review by rinon — APPROVED** [review]

File: src/decode.rs
Comment (rinon) [inline src/decode.rs:5048]: Does this now re-init on each frame? Is that intended?

Comment (kkysen) [reply src/decode.rs:5048]: Hmm, no.  I misread that as the opposite.  Good catch.

Comment (kkysen) [reply src/decode.rs:5048]: In that case, it's probably relying on it being zero-initialized somewhere before, which is not great.  Let me see if there's somewhere we can non-lazily initialize it.

Comment (rinon) [reply src/decode.rs:5048]: Is `Rav1dIntraPredDSPContext` public? We could just add a boolean `initialized` and check that?

**Review by rinon — CHANGES_REQUESTED** [review]
I think we should do something to not reinit on every frame so we don't deviate from the original code.

File: src/decode.rs
Comment (kkysen) [reply src/decode.rs:5048]: No, it's not `pub`.  That'll probably work.  I just want to check where it's actually getting initialized since it's happening implicitly and I want to make sure we avoid UB.

Comment (kkysen) [reply src/decode.rs:5048]: I changed it to use your `initialized: bool` field suggestion.

File: src/decode.rs
Comment (rinon) [inline src/decode.rs:5051]: Is there a reason this should be in the ipred structure rather than the top-level DSP context? We are initializing all of it afaict?

File: src/ipred.rs
Comment (rinon) [inline src/ipred.rs:163]: See other comment, but should this move to the `DSPContext`?

File: src/decode.rs
Comment (kkysen) [reply src/decode.rs:5051]: Hmm, yeah, that's probably better.  I was just confused since the C was checking `ipred`, but that was probably just done randomly.  I'll move it.

File: src/ipred.rs
Comment (kkysen) [reply src/ipred.rs]: Changed in [7d0c2ce..4609590](https://github.com/memorysafety/rav1d/compare/7d0c2ce6da7544f7a9f026d1b44478f4ac19b5ff..4609590678b588824c7770f5d05ba378e2914709).

File: src/decode.rs
Comment (kkysen) [reply src/decode.rs]: Changed in [7d0c2ce..4609590](https://github.com/memorysafety/rav1d/compare/7d0c2ce6da7544f7a9f026d1b44478f4ac19b5ff..4609590678b588824c7770f5d05ba378e2914709).

**Review by randomPoison — APPROVED** [review]
The changes here look fine to me, but make sure you've addressed @rinon's requests before merging.

**Comment by kkysen** [comment]
> The changes here look fine to me, but make sure you've addressed @rinon's requests before merging.

I addressed the last of @rinon's comments and was just waiting on approval, but it was a pretty minimal fix, so I'll go ahead and merge now.

**Comment by rinon** [comment]
I just hadn't gotten a chance to look at it, LGTM thanks 👍

**Pull request merged**
